kbchat/kvstore.go

```diff
@@ -0,0 +1,227 @@
+	GetEntry(teamName string, namespace string, entryKey string) (keybase1.KVGetResult, error)
+	ListNamespaces(teamName string) (keybase1.KVListNamespaceResult, error)
+	ListEntryKeys(teamName string, namespace string) (keybase1.KVListEntryResult, error)
+}
+
+func (a *API) PutEntry(teamName string, namespace string, entryKey string, entryValue string) (result keybase1.KVPutResult, err error) {
+	return a.PutEntryWithRevision(teamName, namespace, entryKey, entryValue, 0)
+}
+
+func (a *API) PutEntryWithRevision(teamName string, namespace string, entryKey string, entryValue string, revision int) (result keybase1.KVPutResult, err error) {
+
+	type PutArgs struct {
+		Method string `json:"method"`
+		Params struct {
+			Options struct {
+				Team       string `json:"team"`
+				Namespace  string `json:"namespace"`
+				EntryKey   string `json:"entryKey"`
+				EntryValue string `json:"entryValue"`
+				Revision   int    `json:"revision,omitempty"`
+			} `json:"options"`
+		} `json:"params"`
+	}
+
+	args := PutArgs{Method: "put"}
+	args.Params.Options.Team = teamName
+	args.Params.Options.Namespace = namespace
+	args.Params.Options.EntryKey = entryKey
+	args.Params.Options.EntryValue = entryValue
+
+	if revision != 0 {
+		args.Params.Options.Revision = revision
+	}
+
+	apiInput, err := json.Marshal(args)
+	if err != nil {
+		return result, err
+	}
+
+	cmd := a.runOpts.Command("kvstore", "api")
+	cmd.Stdin = strings.NewReader(string(apiInput))
+	bytes, err := cmd.Output()
+	if err != nil {
+		return result, APIError{err}
+	}
+
+	entry := PutEntry{}
+	err = json.Unmarshal(bytes, &entry)
+	if err != nil {
+		return result, UnmarshalError{err}
+	}
+	if entry.Error.Message != "" {
+		return result, entry.Error
+	}
+	return entry.Result, nil
+}
+
+func (a *API) DeleteEntry(teamName string, namespace string, entryKey string) (result keybase1.KVDeleteEntryResult, err error) {
+	return a.DeleteEntryWithRevision(teamName, namespace, entryKey, 0)
+}
+
+func (a *API) DeleteEntryWithRevision(teamName string, namespace string, entryKey string, revision int) (result keybase1.KVDeleteEntryResult, err error) {
+
+	type DeleteArgs struct {
+		Method string `json:"method"`
+		Params struct {
+			Options struct {
+				Team      string `json:"team"`
+				Namespace string `json:"namespace"`
+				EntryKey  string `json:"entryKey"`
+				Revision  int    `json:"revision,omitempty"`
+			} `json:"options"`
+		} `json:"params"`
+	}
+
+	args := DeleteArgs{Method: "del"}
+	args.Params.Options.Team = teamName
+	args.Params.Options.Namespace = namespace
+	args.Params.Options.EntryKey = entryKey
+
+	if revision != 0 {
+		args.Params.Options.Revision = revision
+	}
+
+	apiInput, err := json.Marshal(args)
+	if err != nil {
+		return result, err
+	}
+
+	cmd := a.runOpts.Command("kvstore", "api")
+	cmd.Stdin = strings.NewReader(string(apiInput))
+	bytes, err := cmd.Output()
+	if err != nil {
+		return result, APIError{err}
+	}
+
+	entry := DeleteEntry{}
```

Comment:
so, i get that you're defining `DeleteArgs` inside this method because it's totally internal to right here. but so is this type. why is this one defined above, but `DeleteArgs` defined inside the method? 

i would suggest defining both of these outside of the method, and then you can signal that they're private by calling them something like `kvstoreDeleteRequest` and `kvstoreDeleteResponse`.

Reply:
and same for all the others

Reply:
you could actually cover all of these request apis with a single set of types. 
```golang
type kvstoreMethod string
type kvstoreOptions struct {
	Team       string  `json:"team"`
	Namespace  *string `json:"namespace,omitempty"`
	EntryKey   *string `json:"entryKey,omitempty"`
	EntryValue *string `json:"entryValue,omitempty"`
	Revision   *int    `json:"revision,omitempty"`
}
type kvstoreParams struct {
	Options kvstoreOptions `json:"options"`
}
type kvstoreAPIReq struct {
	Method kvstoreMethod `json:"method"`
	Params kvstoreParams `json:"params"`
}
```

Reply:
good idea!